js/style/style.js

```diff
@@ -172,13 +172,9 @@ Style.prototype = util.inherit(Evented, {
     },
 
     _broadcastLayers: function() {
-        var ordered = [];
-
-        for (var id in this._layers) {
-            ordered.push(this._layers[id].json());
-        }
```

Comment:
While not in the official spec, all major JS interpreters will iterate over objects in the order that the keys were added. 

If a call to `addLayer` specifies a `before` parameter, it'll end up at the end of this array, even though it was intended for elsewhere. Then, during label placement in the worker, the layer will not be ordered correctly.